In [1]:
# Import Dependencies
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split Data into Training and Testing Sets ##

#### Step 1: Read `lending_data.csv` from `Resources` folder into Pandas DataFrame ####

In [2]:
# Read 'lending_data.csv' File From Resources Folder Into Pandas DataFrame
lending_data = Path('Resources/lending_data.csv')
lending_data_df = pd.read_csv(lending_data) 

# Review DataFrame
lending_data_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

#### Step 2: Create labels set (`y`)  from “loan_status” column & features (`X`) DataFrame from remaining columns ####

In [3]:
# Separate y Variable, The Labels
y = lending_data_df['loan_status']

# Separate X Variable, The Features
X = lending_data_df.drop(columns='loan_status')

In [4]:
# Review y Variable Series
y[-5:]

77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, dtype: int64

In [5]:
# Review X Variable DataFrame
X[:5]

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

#### Step 3: Split data into training and testing datasets by using `train_test_split` ####

In [6]:
# Import train_test_learn Module
from sklearn.model_selection import train_test_split

# Split Data Using train_test_split, Assign random_state Of 1 To The Function
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=1, 
                                                    stratify=y)

---

## Create Logistic Regression Model with the Original Data ##

####  Step 1: Fit logistic regression model by using training data (`X_train` and `y_train`) ####

In [7]:
# Import LogisticRegression Module From SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate Logistic Regression Model, Assign random_state Parameter of 1 To The Model
logistic_regression_model = LogisticRegression(random_state=1)

# Fit The Model Using Training Data
logistic_regression_model.fit(X_train, y_train)

LogisticRegression(random_state=1)

#### Step 2: Save predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model ####

In [8]:
# Make Prediction Using The Testing Data
testing_predictions = logistic_regression_model.predict(X_test)

#### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [9]:
# Generate Confusion Matrix For The Model
testing_confusion_matrix = confusion_matrix(y_test, testing_predictions)
testing_confusion_matrix_df = pd.DataFrame(testing_confusion_matrix,
                                           index=['Healthy Loans (Actual)',
                                                  'High-Risk Loans (Actual)'],
                                           columns=['Healthy Loans (Predicted)',
                                                    'High-Risk Loans (Predicted)'])
testing_confusion_matrix_df

Healthy Loans (Predicted)  \
Healthy Loans (Actual)                        18679   
High-Risk Loans (Actual)                         67   

                          High-Risk Loans (Predicted)  
Healthy Loans (Actual)                             80  
High-Risk Loans (Actual)                          558

In [11]:
# Print Classification Report For The Model
testing_classification_report = classification_report(y_test,
                                                      testing_predictions,
                                                      target_names=('Healthy Loans (0)',
                                                                    'High-Risk Loans (1)'))
print(testing_classification_report)

                     precision    recall  f1-score   support

  Healthy Loans (0)       1.00      1.00      1.00     18759
High-Risk Loans (1)       0.87      0.89      0.88       625

           accuracy                           0.99     19384
          macro avg       0.94      0.94      0.94     19384
       weighted avg       0.99      0.99      0.99     19384



#### Step 4: Answer the following question ####

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 

Based on the Classification Report:
  1. Precision:
      - For healthy loans (label 0), the precision is 1.00, indicating that when the model predicts a loan as healthy, it is correct 100% of the time.
      - For high-risk loans (label 1), the precision is 0.87, meaning that when the model predicts a loan as high-risk, it is correct about 87% of the time.
  2. Recall:
      - For healthy loans, the recall is 1.00, indicating that the model correctly identifies all healthy loans in the dataset.
      - For high-risk loans, the recall is 0.89, suggesting that the model captures about 89% of the actual high-risk loans.
  3. F1-Score:
      - For healthy loans, the F1-score is 1.00, indicating a perfect balance between precision and recall
      - For high-risk loans, the F1-score is 0.88, reflecting a slightly lower balance between precision and recall compared to healthy loans.
  4. Support:
      - There are 18759 healthy loans and 625 high-risk loans in the dataset.

Based on the Confusion Matrix:
  - Out of 18759 actual healthy loans, 18679 are correctly predicted as healthy, and only 80 are incorrectly predicted as high-risk.
  - Out of 625 actual high-risk loans, 558 are correctly predicted as high-risk, and only 67 are incorrectly predicted as healthy.

In summary:
  - The logistic regression model performs well in predicting both healthy (0) and high-risk loans (1). It has high precision, recall, and F1-score for both classes, indicating good overall performance.

---